### ISSUE TITLE
Problems with qualitative variables with one modality

### ISSUE THREAD

**Issue by ThomasDeroyon** [martinchevalier/gustave#8]
The variance wrappers created with gustave have difficulties with qualitative variables with one modality. This problem can also occur when using the by or where parameters in case the qualitative variable of interest has only one modality in the modality of the where variable, or in one modality of the by variable :
```

# Simple variance function

compute_variance <- function(y){
  
  rep(1, ncol(y))
  
}

# Data generation

set.seed(215)

d <- data.frame(id = 1 : 1000, 
                weight = c(rep(10, 1000)),
                x = 1000 * rnorm(1000, mean = 10, sd = 2), 
                a = c(rep('a', 500), rep('b', 500)),
                y = rep('c', 1000),
                z = c(rep('m1', 250), rep('m2', 250), rep('m3', 500))
                )

# Variance wrapper definition

precision <- 
  gustave::define_variance_wrapper(variance_function = compute_variance, 
                                   reference_id = d$id, 
                                   reference_weight = d$weight, 
                                   default_id = 'id')

# The wrapper is working :

precision(d, mean(x))

# But not with one modality qualitative variables :
 
precision(d, mean(y))
precision(d, mean(z), by = a)
precision(d, mean(z), where = (a == 'b'))

```

The problem comes from inner function discretize_qualitative_var, which cannot handel qualitative variables with one modality. It is sufficient to add a default behaviour in case the qualitative variable discretize_qualitative_var is applied to has only one modality, and to modify the version of discretize_qualitative_var in the parent environements of the precision function to correct the problem : 

```
discretize<- function (var, logical = FALSE){
  
  var <- droplevels(as.factor(var))
  result <- Matrix(nrow = length(var), ncol = length(levels(var)))
  
  if (length(levels(var)) == 1){  
    
    result[!is.na(var), ] <- 1
    
    
    } else {
      
      result[!is.na(var), ] <- Matrix::sparse.model.matrix(~var - 
                                                         1)
      
      }
  
  result[is.na(var), ] <- NA
  
  if (!logical) result <- result * 1
  
  
  rownames(result) <- names(var)
  colnames(result) <- levels(var)
  result
  
  
  }

parent.env(environment(precision))$discretize_qualitative_var <- discretize

environment(parent.env(environment(precision))$total)$discretize_qualitative_var <- discretize
parent.env(environment(parent.env(environment(precision))$total))$discretize_qualitative_var <- discretize

environment(parent.env(environment(precision))$mean)$discretize_qualitative_var <- discretize
parent.env(environment(parent.env(environment(precision))$mean))$discretize_qualitative_var <- discretize

precision(d, mean(y))

precision(d, mean(z), where = a == 'b')

precision(d, mean(z), by = a)

```

**Comment by martinchevalier** [comment]
Hi Thomas, 

And thank you for this issue and the suggested bug fix. I try to look at it in the forthcoming days. 

Martin

**Comment by martinchevalier** [comment]
Fixed with b1de6181f35e907db734c051b2f5b1f8a49a06e9

@ThomasDeroyon, if you install the parckage from github (with remotes::install_github()) and re-generate your variance wrapper, it should fix the problem you raised. 

Thanks again for raising this issue and for the fix (I introduced it as is in the source code, including your MWE).
